dialogue-core/src/main/java/com/palantir/dialogue/core/RetryingChannel.java

```diff
@@ -235,10 +228,13 @@ private final class RetryingCallback {
         private final DetachedSpan span = DetachedSpan.start("Dialogue-RetryingChannel");
         private int failures = 0;
 
+        private Optional<RequestBodyWrapper> maybeWrappedRequestBody;
+
         private RetryingCallback(
                 Endpoint endpoint, Request request, Optional<SafeRuntimeException> callsiteStacktrace) {
             this.endpoint = endpoint;
             this.request = request;
+            this.maybeWrappedRequestBody = request.body().map(RequestBodyWrapper::new);
```

Comment:
I think you need to actually change request to RequestBodyWrapper so that's the body you're actually using for the writes? and then consumed needs to be volatile at least?

Reply:
> I think you need to actually change request to RequestBodyWrapper so that's the body you're actually using for the writes?
Agree. I think this is what Carter was also getting at with the pointer to `ContentEncodingChannel`.

> and then consumed needs to be volatile at least?
Missed this. Makes sense as we'd like the writes to the `consumed` field complete before any `read` from this retrying logic.